tools/autograd/gen_variable_type.py

```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
         f.write(template.substitute(env))
 
 
-def load_derivatives(path):
-    with open(path, 'r') as f:
-        definitions = yaml.load(f, Loader=Loader)
+def saved_variables(formula, args):
+    # find which arguments need to be saved
+    saved = []
 
-    # Matches "foo" in "foo, bar" but not "foobar". The name is substituted for
-    # the {} characters.
-    name_regex = r'(^|\W){}($|\W)'
+    for arg in args:
+        if 'name' not in arg:
+            # some returned arguments do not have names
+            continue
+        name = arg['name']
+
+        def replace_sizes(m):
+            res = name + '_sizes'
+            saved.append({'name': res, 'type': 'IntList'})
+            return res
+
+        def replace_zeros(m):
+            r = name + '_info'
+            saved.append({'name': r, 'type': 'TypeAndSize'})
+            return r + '.zeros()'
+
+        def replace_size_n(m):
+            res = name + '_argsize_{}'.format(*m.groups())
+            saved.append({'name': res, 'type': 'int64_t'})
+            return res
+
+        def replace_to_arg_sizes(m):
+            res = name + '_argsizes_{}'.format(*m.groups())
+            saved.append({'name': res, 'type': 'IntList'})
+            return res
+
+        # replace self.sizes() with self_sizes
+        formula = re.sub(r'{}.sizes\(\)'.format(name), replace_sizes, formula)
+        # replace zeros_like(self) with self_info
+        formula = re.sub(r'zeros_like\({}\)'.format(name), replace_zeros, formula)
+        # replace self.size(2) with self_size_2
+        formula = re.sub(r'{}.size\((\w+)\)'.format(name), replace_size_n, formula)
+        # replace to_arg_sizes(self, 2) with self_argsizes_2
+        formula = re.sub(r'to_arg_sizes\({}, (\w+)\)'.format(name), replace_to_arg_sizes, formula)
+
+        if re.search(IDENT_REGEX.format(name), formula):
+            arg = copy.deepcopy(arg)
+            arg['type'] = arg['type'].replace('const ', '').replace(' &', '')
+            saved.append(arg)
+    return formula, saved
+
+
+def create_derivative(declaration, formula, output_indices, var_names):
+    returns = [r for r in declaration['returns'] if r.get('name') != 'self']
+    arguments = [arg for arg in declaration['arguments']]
+    if any(arg['name'] == 'inplace' for arg in arguments):
+        for arg in arguments:
+            if arg['name'] == 'input':
+                returns += [arg]
```

Comment:
this logic is a bit weird. If we find `inplace` in argument list, we look for `input` and add it to returned values? Is it empty before or what?

Reply:
This is for NN functions. For something like: `relu(Tensor input, bool inplace)`. If inplace is true then `input` is also a return argument. It's important that if we save input for the backwards we do it in the `save_outputs` section not the `save_inputs` because it might be modified.